decoder/src/lib.rs

```diff
@@ -382,6 +382,84 @@ impl<'t, 'b> Decoder<'t, 'b> {
         Ok(())
     }
 
+    /// Sort and deduplicate `params` so that they can be interpreted correctly during decoding
+    fn prepare_params(&self, params: &mut Vec<Parameter>) {
+        // sort & dedup to ensure that format string args can be addressed by index too
+        params.sort_by(|a, b| {
+            if a.index == b.index {
+                match (&a.ty, &b.ty) {
+                    (Type::BitField(a_range), Type::BitField(b_range)) => {
+                        a_range.start.cmp(&b_range.start)
+                    }
+                    _ => Ordering::Equal,
+                }
+            } else {
+                a.index.cmp(&b.index)
+            }
+        });
+
+        // deduplicate bitfields by merging them into a new one with range min..max
+        // TODO refactor when `drain_filter()` is stable
+        // sorry about the wonky vars but accessing enum fields is too messy to just use a Param{}
+        let mut curr_bitfield_range: Option<Range<u8>> = None;
+        let mut curr_bitfield_index = 0;
+        let mut i = 0;
+        let initial_num_params = params.len();
+        let mut num_params_read = 0;
+        while num_params_read < initial_num_params {
+            match &(&mut params[i]).ty {
```

Comment:
Any reason this doesn't work?

```suggestion
            match &params[i].ty {
```